### PR TITLE
added classes parameter to extract-region-processor; enables extracti…

### PR DESCRIPTION
**Pull request by JBBalling**
…on of specific regions e.g. TableRegion

### CODE REVIEW

File: ocrd_segment/extract_regions.py [hunk 1]
```diff
@@ -86,7 +100,7 @@ def process(self):
             ptype = page.get_type()
 
             regions = dict()
-            for name in CLASSES.keys():
+            for name in classes.keys():
                 if not name or name == 'Border' or ':' in name:
```
Comment (bertsky) [inline ocrd_segment/extract_regions.py]: I'm afraid this stopped working after I added the sub-region types to `CLASSES` for the new drawing feature of extract-pages. – This is independent of your addition.
```suggestion
                if not name or not name.endswith("Region"):
```

File: ocrd_segment/extract_regions.py [hunk 2]
```diff
@@ -66,6 +66,20 @@ def process(self):
...
+                provided_classes = self.parameter["classes"].split(",")
+            else:
+                provided_classes = self.parameter["classes"]
+
+            if isinstance(provided_classes, list):
+                for regionClass in provided_classes:
+                    classes = {regionClass: classes[regionClass]}
+            else:
+                # only 1 class provided
+                classes = {provided_classes: classes[provided_classes]}
```
Comment (bertsky) [inline ocrd_segment/extract_regions.py]:
```suggestion
            selected_classes = self.parameter["classes"].split(",")
            classes = { region: classes[region] for region in selected_classes }
```

File: ocrd_segment/ocrd-tool.json [hunk 3]
```diff
@@ -291,6 +291,27 @@
...
+            "MusicRegion",
+            "GraphicRegion",
+            "UnknownRegion",
+            "CustomRegion",
+            "SeparatorRegion",
+            "MathsRegion",
+            "TextRegion",
+            "MapRegion",
+            "ChemRegion",
+            "TableRegion"
```
Comment (bertsky) [inline ocrd_segment/ocrd-tool.json]: I'm afraid this does not work (and is not valid). With `enum` you enforce any if these values, but thereby also prohibit combining them with commas.

If you want to enforce this syntactically, you'd have to set `type=array`, and then place an `enum` on its `items`. (But then, of course, the code would have to be adapted as well (by removing comma-splitting).

```suggestion
```

File: ocrd_segment/ocrd-tool.json [hunk 4]
```diff
@@ -291,6 +291,27 @@
...
+          ],
+          "description": "Comma-separated string of classes to extract e.g. 'TextRegion, TableRegion, ImageRegion'"
```
Comment (bertsky) [inline ocrd_segment/ocrd-tool.json]:
```suggestion
          "description": "Comma-separated string of region types to extract e.g. 'TextRegion,TableRegion,ImageRegion'. If empty, all regions are allowed."
```

**Review by bertsky — CHANGES_REQUESTED** [review]
Great idea!

File: ocrd_segment/ocrd-tool.json [hunk 5]
```diff
@@ -291,6 +291,27 @@
           "default": "",
           "description": "Comma-separated list of forbidden image features (e.g. `binarized,despeckled`)."
         },
+        "classes": {
+          "type": "string",
+          "enum": [
```
Comment (kba) [inline ocrd_segment/ocrd-tool.json]:
```suggestion
          "pattern": "((Noise|LineDrawing|Advert|Image|Chart|Music|Graphic|Unknown|Custom|Separator|Maths|Text|Map|Chem|Table)Region(,(Noise|LineDrawing|Advert|Image|Chart|Music|Graphic|Unknown|Custom|Separator|Maths|Text|Map|Chem|Table)Region)*)?" 
```

:)

Comment (bertsky) [reply ocrd_segment/ocrd-tool.json]: Oh, fascinating! Does the JSON parser really enforce this?

Comment (kba) [reply ocrd_segment/ocrd-tool.json]: Yes, `pattern` is generally enforced. Whether it's sensible is another question, the resulting error message won't be very intuitive for the user.

**Review by bertsky — APPROVED** [review]
Excellent, thanks!

**Pull request merged**